state/accumulator.go

```diff
@@ -153,8 +170,26 @@ func (a *Accumulator) Initialise(roomID string, state []json.RawMessage) (bool,
 			return fmt.Errorf("error fetching snapshot id for room %s: %s", roomID, err)
 		}
 		if snapshotID > 0 {
-			// we only initialise rooms once
-			logger.Info().Str("room_id", roomID).Int64("snapshot_id", snapshotID).Msg("Accumulator.Initialise called but current snapshot already exists, bailing early")
+			// Poller A has received a gappy sync v2 response with a state block, and
+			// we have seen this room before. If we knew for certain that there is some
+			// other active poller B in this room then we could safely skip this logic.
+
+			// Log at debug for now. If we find an unknown event, we'll return it so
+			// that the poller can log a warning.
+			logger.Debug().Str("room_id", roomID).Int64("snapshot_id", snapshotID).Msg("Accumulator.Initialise called with incremental state but current snapshot already exists.")
+			eventIDs := make([]string, len(state))
+			for i := range state {
+				eventIDs[i] = gjson.ParseBytes(state[i]).Get("event_id").Str
+			}
+			unknownEventIDs, err := a.eventsTable.SelectUnknownEventIDs(txn, eventIDs)
+			if err != nil {
+				return fmt.Errorf("error determing which event IDs are unknown: %s", err)
+			}
+			for i := range state {
+				if _, unknown := unknownEventIDs[eventIDs[i]]; unknown {
```

Comment:
Shape of this is a bit weird. We could just make a stateMap before and then just do lookups, which would avoid the O(n) here. In the first O(n) when pulling out the `event_id`, make a map with event ID to `[]byte`, then use that here.

Reply:
>  I assumed that the state block was ordered with the earliest event first and wanted to preserve this order. 

AFAICS Synapse will chuck out the contents of the state block in any random order, so I'm happy to make the change you suggest.